hail/src/main/scala/is/hail/expr/ir/EmitStream.scala

```diff
@@ -8,9 +8,399 @@ import is.hail.expr.types.physical._
+    def apply(eos: Code[Ctrl], push: B => Code[Ctrl])(implicit ctx: EmitStreamContext): Source[B] = {
+      val source = stream(
+        eos = close.map(Code(_, eos)).getOrElse(eos),
+        push = f(ctx, _, b => push(b)))
+      Source[B](
+        setup0 = setup0.map(Code(_, source.setup0)).getOrElse(source.setup0),
+        close0 = close0.map(Code(_, source.close0)).getOrElse(source.close0),
+        setup = setup.map(Code(_, source.setup)).getOrElse(source.setup),
+        close = close.map(Code(_, source.close)).getOrElse(source.close),
+        pull = source.pull)
+    }
+  }
+
+  def map[A, B](stream: Stream[A])(
+    f: A => B,
+    setup0: Option[Code[Unit]] = None,
+    setup:  Option[Code[Unit]] = None,
+    close0: Option[Code[Unit]] = None,
+    close:  Option[Code[Unit]] = None
+  ): Stream[B] = mapCPS(stream)((_, a, k) => k(f(a)), setup0, setup, close0, close)
+
+  def addFinalizer[A](stream: Stream[A])(finalize: Code[Unit]): Stream[A] =
+    map[A, A](stream)(a => a, close = Some(finalize))
+
+  def flatMap[A](outer: Stream[Stream[A]]): Stream[A] = new Stream[A] {
+    def apply(eos: Code[Ctrl], push: A => Code[Ctrl])(implicit ctx: EmitStreamContext): Source[A] = {
+      val outerPullJP = joinPoint()
+      var innerSource: Source[A] = null
+      val innerPullJP = joinPoint()
+      val hasBeenPulled = newLocal[Code[Boolean]]
+      val outerSource = outer(
+        eos = eos,
+        push = inner => {
+          innerSource = inner(
+            eos = outerPullJP(()),
+            push = push)
+          innerPullJP.define(_ => innerSource.pull)
+          Code(innerSource.setup, innerPullJP(()))
+        })
+      outerPullJP.define(_ => outerSource.pull)
+      Source[A](
+        setup0 = Code(hasBeenPulled := false, outerSource.setup0, innerSource.setup0),
+        close0 = Code(innerSource.close0, outerSource.close0),
+        setup = Code(hasBeenPulled := false, outerSource.setup),
+        close = Code(hasBeenPulled.load.mux(innerSource.close, Code._empty),
+                     outerSource.close),
+        pull = hasBeenPulled.load.mux(innerPullJP(()),
+                                      Code(hasBeenPulled := true, outerPullJP(()))))
+    }
+  }
+
+  def filter[A](stream: Stream[COption[A]]): Stream[A] = new Stream[A] {
+    def apply(eos: Code[Ctrl], push: A => Code[Ctrl])(implicit ctx: EmitStreamContext): Source[A] = {
+      val pullJP = joinPoint()
+      val source = stream(
+        eos = eos,
+        push = _.apply(none = pullJP(()), some = push))
+      pullJP.define(_ => source.pull)
+      source.copy(pull = pullJP(()))
+    }
+  }
+
+  def filter[A: ParameterPack](stream: Stream[A], cond: A => Code[Boolean]): Stream[A] =
+    filter(mapCPS[A, COption[A]](stream)((_ctx, a, k) => {
+      implicit val ctx = _ctx
+      val as = newLocal[A]
+      Code(as := a, k(COption(!cond(as.load), as.load)))
+    }))
+
+  def zip[A, B](left: Stream[A], right: Stream[B]): Stream[(A, B)] = new Stream[(A, B)] {
```

Comment:
Previously we discussed streams becoming streams of, Region => EmitTriplet, I guess.  I want to make sure we're on the same page about what that means.  In part this is in relation to my PR from yesterday which begins to introduce PValue into the emitter: https://github.com/hail-is/hail/pullMake EmitTriplet value a PValue/8144 Make EmitTriplet value a PValue.  I'm thinking about how this will all fit together.

The point of stream is to compile the stream nodes.  The functions you've implemented look to my like cutting the stream node implementations effectively in half at types that I/we don't normally think of being values inside our compiled language (Scala tuples, indexed seq, etc.)  One option is to add them as you've done with ParameterPack.  Alternatively, these stream functions can roughly correspond to the nodes themselves.  Unless there a significant advantage to the former, I prefer the latter.  The main advantage is that then values in streams (and everywhere) will be typed by our type system.  It will let us remove ParameterPack and I think make the implementations more transparent.

Thoughts?  This can again go on the todo list for this stream (ha) of work.

Reply:
> The point of stream is to compile the stream nodes.

Agreed. I see `Stream` and `Source` as being private to the emit implementations of the stream nodes. I don't know yet what the final code organization will be (should stream nodes be lumped in with the rest in `Emit`?), but I'll want to make the visibility of these stream abstractions as small as possible.

Code generation for streams is hard, so my main goal with "cutting the stream node implementations in half" was to separate the concerns of generating the right control flow from things like missingness logic and region management, as well as to share code between different stream nodes. Since this is all localized to just emit for stream nodes, it should be easy enough to refactor when its all working if that separation isn't helpful.

> The functions you've implemented look to my like cutting the stream node implementations effectively in half at types that I/we don't normally think of being values inside our compiled language (Scala tuples, indexed seq, etc.) One option is to add them as you've done with ParameterPack.

The causality went the other way. Milo added `ParameterPack` in his stream work, to support zip without allocating tuples (and in the old design it was also needed to support streams with compound state without allocating). I'm using scala tuples and indexed seqs because that's how `ParameterPack` is designed, but I don't see any obstruction to replacing all that with PValues.

> The main advantage is that then values in streams (and everywhere) will be typed by our type system. It will let us remove ParameterPack and I think make the implementations more transparent.

I agree.

Reply:
> I see Stream and Source as being private to the emit implementations of the stream nodes.

Great, I have the same picture.

 > I'm using scala tuples and indexed seqs because that's how ParameterPack is designed, but I don't see any obstruction to replacing all that with PValues.

Great, although I don't see PValues as necessarily replacing, for example, the IndexedSeq[blah] in multiZip.  In my picture, the PValues should correspond to values realized by the IR being compiled.  The IndexedSeq in multiZip isn't a value that actually occurs in the program, it stores the values for the zipped elements that get added to the environment for the evaluation of the body.  So the IndexedSeq[blah] would stay as an intermediate value in the compilation of multiZip which gets fed to something like env.bindIterables.

Reply:
Right, makes sense. What you described about the IndexedSeq is how I think of the element type in `Stream[A]` in general: it isn't a value that actually occurs in the program, it's an intermediate value in the compilation which gets fed to some other piece of the compiler.

Reply:
Right, I get that.  The threads I'm thinking about here:

The question is, what does Stream[A] need to know about A to work?  I understand you can do that (ParameterPack), but we're building a lot of infrastructure elsewhere in the backend for working with an manipulating values and if you cut along those values, you get what you need for free, and if you don't, you have to build some parallel infrastructure that we have to maintain.

I also don't want pieces of the backend to become little islands of their own, incompatible infrastructure.  A bit of that is to be expected, of course, but when we could leverage common infrastructure, every part benefits and it becomes easier to work across the whole codebase.  I feel like that happened a little bit with the first stream implementation, and my instinct is to push back on it.  (The ParamterPack/TypedEmit was the hardest part about the PValue refactor, for example.)

I guess I see three outcomes:

ParameterPack is the right thing for Stream and it stays.

We cut along EmitTriplets, etc. and ParameterPack goes.

ParameterPack for is the right thing for a bunch of stuff and we lift it into the rest of the code.

In general, I prefer one of the latter two outcomes if possible (which I think is what happened in our discussion of TypedTriplet).

For the operations you have so far, the intermediate values in Stream[A] seem mostly ones that will be placed into bindings.  The other answer might be we need cleaner abstractions for managing bindings that you can take advantage of, which will make it more attractive to cut along EmitTriplets.  Binding has two parts: memoizing and updating the env, but I think of them together.  Your code splits out the just the memoization.

Reply:
> The question is, what does Stream[A] need to know about A to work?

I think the answer is: nothing. It's a complete black box that the stream just needs to pass to its consumer. My primary reason for looking for a design like that is to make grouping operators easier. Grouping operators can return a `Stream[Stream[A]]`, followed by something like `.map(_.aggregate)`, and it doesn't matter that `Stream[A]` isn't something that can be bound in an environment.

Only a few stream combinators require `ParameterPack`, like fold and unfold to maintain state, and mux to have a single consumer label accept values from two possible sources (by pushing the value on the stack before jumping).

> we're building a lot of infrastructure elsewhere in the backend for working with an manipulating values and if you cut along those values, you get what you need for free, and if you don't, you have to build some parallel infrastructure that we have to maintain.

I think the surface area of stream code that would need to change if we change value manipulating infrastructure is quite small. I see that as a benefit of using scala types like tuples.

Reply:
> I think the answer is: nothing.

OK, good, I can get behind that.  I see now that some operations don't require ParameterPack.  And we can switch the state to to some Emit/PValue representation which feels natural anyway (or just make the state implicit).

 > I think the surface area of stream code that would need to change if we change value manipulating infrastructure is quite small. I see that as a benefit of using scala types like tuples.

OK, I think we're in agreement.  I like that Stream[A] has an arbitrary A and you can put whatever you want in there.  Very nice.  I think intertwining A with the backend value stack is the mistake: it should either be the backend value stack, or decoupled form it.  ParameterPack is doing the intertwining, and again I think that is my main issue.

The question remains, what to do with mux.  I have a proposal: mux should call the incoming push twice.  The caller can then be responsible for managing the values.

Reply:
> The question remains, what to do with mux. I have a proposal: mux should call the incoming push twice. The caller can then be responsible for managing the values.

I'm not sure I follow. I think this is what it does now. The consumer's push code gets a label, and mux jumps to it from two different places. If you're objecting to passing the value on the operand stack, it could just as easily store it in a local.

Reply:
> I'm not sure I follow. I think this is what it does now. The consumer's push code gets a label, and mux jumps to it from two different places. If you're objecting to passing the value on the operand stack, it could just as easily store it in a local.

I'm objecting to A being considered value-like.  I'm saying rather than assume stuff about A, literally call the push function twice with the two As for the two different streams and don't manipulate the stack, locals or values.  The consumer will be responsible for manipulating the values and jumping to a common place to avoid code duplication (if necessary).

Or just move it to the implementation of If in EmitStream2 and use the emitter value representations.

Reply:
Okay, I get it now.

> Or just move it to the implementation of If in EmitStream2 and use the emitter value representations.

Yeah, I like that.